index.js

```diff
@@ -177,21 +181,71 @@ program
 */
 
 
-var ercChecker = function (originalHTML, reproducedHTML, outputPath, checkID, createParentDirectoriesInOutputPath, silenceDebuggers, saveMetadata) {
+var ercChecker = function (originalHTML, reproducedHTML, checkID, outputPath, saveDiffHTML, saveMetadataJSON, createParentDirectoriesInOutputPath, silenceDebuggers) {
 
+	var checkId = '';
 	return new Promise( function (resolve, reject) {
 
-
-
 		var pathOriginalHTML = originalHTML,
 			pathReproducedHTML = reproducedHTML,
 			brokenPath = false,
 			checkStart = Date.now();
 
-		var outputName = outputPath;
+		if ((saveDiffHTML || saveMetadataJSON) && outputPath == null) {
+			debugERROR("Cannot save files without output path.".red);
+			reject(new Metadata(checkStart, new Error("Requested saving check files without specifying an output path.")));
+		}
+
+		debug(checkID);
+
 		if (checkID) {
-			var checkIDString = '_'+checkID;
+			checkId = '_'+checkID;
+		}else {
+			checkId = '';
 		}
+		var fileOutputPath;
+		try {
+			if (outputPath) {
+				fileOutputPath = fs.realpathSync(outputPath);
+			}
+		}
+		catch (e) {
+			if (createParentDirectoriesInOutputPath) {
+
+				if (process.platform === 'win32') {
+					exec("mkdir " + fileOutputPath,
+						function (err) {
+							if (err) {
+								debugERROR("Could not create output directory.".red);
+								reject( new Metadata(checkStart, err) );
+							}
+						}
+					);
+				}
+				else {
+					exec("mkdir -p " + fileOutputPath,
+						function (err) {
+							if (err) {
+								debugERROR("Could not create output directory.".red);
+								reject( new Metadata(checkStart, err) );
+							}
+						}
+					);
+				}
+				fileOutputPath = fs.realpathSync(outputPath);
+			}
+			else {
+				debugERROR("Output path non-existent, option to create it was not set. ".red);
+				reject( new Metadata(checkStart, "Output path non-existent, option to create it was not set. " + e))
```

Comment:
If you reject something, why not reject it with an `Error`? A class `Metadata` by name does not look like it is supposed to contain error messages.

Reply:
The tool, in any case, reject or resolve, returns a Metadata object, containing info on the entire checking job until the point where it breaks.
The idea was to not loose the data collected so far. Any rejected Promise means that the Check went wrong, and the Rejection parameter contains an object that holds the encountered Error(s). There should not be errors in a Resolve parameter Metadata object.
 
However, Errors may only occur when trying to save files to an invalid path. In this case,  any progress would be lost. My idea was to keep this progress as the Metadata object, so we store the start and end dates of the check, and maybe get more hints as to what went wrong in troubleshooting. 

The Metadata object could as well be split into two classes, one for Error-free checks, and one for rejections. Or we only return Errors.